lib/ansible/plugins/action/set_fact.py

```diff
@@ -37,25 +38,29 @@ def run(self, tmp=None, task_vars=None):
         result = super(ActionModule, self).run(tmp, task_vars)
         del tmp  # tmp no longer has any effect
 
-        facts = dict()
-
+        facts = {}
         cacheable = boolean(self._task.args.pop('cacheable', False))
 
         if self._task.args:
             for (k, v) in iteritems(self._task.args):
                 k = self._templar.template(k)
 
                 if not isidentifier(k):
-                    result['failed'] = True
-                    result['msg'] = ("The variable name '%s' is not valid. Variables must start with a letter or underscore character, and contain only "
-                                     "letters, numbers and underscores." % k)
-                    return result
+                    raise AnsibleActionFail("The variable name '%s' is not valid. Variables must start with a letter or underscore character, "
+                                            "and contain only letters, numbers and underscores." % k)
 
+                #NOTE: this should really use BOOLEANS from convert_bool, but only in the k=v case, right now it converts matching explicit yaml strings also.
                 if not C.DEFAULT_JINJA2_NATIVE and isinstance(v, string_types) and v.lower() in ('true', 'false', 'yes', 'no'):
```

Comment:
Doesn't checking DEFAULT_JINJA2_NATIVE first address that though?

Reply:
only when native, when 'non native' it still makes these 2 the same:
```
normallybool: no
normallystring: "no"
```

Reply:
when non native, it seems like they both become booleans anyway

Reply:
yes, that is my point, they really should not, but not changing implementation at this point, why i added note, clarified a bit more now

Reply:
Oh, I misread your comment too